xmtp_mls/src/subscriptions/stream_conversations.rs

```diff
@@ -131,6 +131,86 @@ where
     }
 }
 
+pin_project! {
+    /// A stream that polls two inner streams.
+    /// this stream closes once either one of the streams close.
+    /// gives priority to S1 (S1 will always be polled first.)
+    /// If S1 finishes before S2, will attempt to drain S2 of items.
+    pub struct MultiplexedStream<S1, S2, T> {
+        #[pin] s1: S1,
+        #[pin] s2: S2,
+        queue: VecDeque<T>,
+        terminated: bool,
+    }
+}
+
+impl<S1, S2, T> MultiplexedStream<S1, S2, T> {
+    pub fn new(s1: S1, s2: S2) -> Self {
+        Self {
+            s1,
+            s2,
+            queue: Default::default(),
+            terminated: false,
+        }
+    }
+}
+
+// TODO: unit test this
+impl<S1, S2, T> Stream for MultiplexedStream<S1, S2, T>
+where
+    S1: Stream<Item = T>,
+    S2: Stream<Item = T>,
+{
+    type Item = T;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.as_mut().project();
+        if *this.terminated {
+            if let Some(item) = this.queue.pop_front() {
+                if !self.queue.is_empty() {
+                    cx.waker().wake_by_ref();
+                }
+                return Poll::Ready(Some(item));
+            }
+            return Poll::Ready(None);
+        }
+        match this.s1.as_mut().poll_next(cx) {
+            Poll::Ready(Some(item)) => {
+                return Poll::Ready(Some(item));
+            }
+            Poll::Ready(None) => {
+                *this.terminated = true;
+                while let Poll::Ready(Some(item)) = this.s2.as_mut().poll_next(cx) {
+                    this.queue.push_back(item);
+                }
+                if this.queue.is_empty() {
+                    return Poll::Ready(None);
+                } else {
+                    cx.waker().wake_by_ref();
+                    return Poll::Pending;
+                }
+            }
+            Poll::Pending => (),
+        };
+
+        match this.s2.as_mut().poll_next(cx) {
```

Comment:
Why is the business logic simpler for s2 than s1?

Reply:
Since S1 is polled first, S2 can be simpler -- S1 would have returned Poll::Ready with an item before ever reaching S2. Therefore, we can assume that if we got to S2, S1 is pending and doesn't have anything sitting in its queue

Whereas if S1 finishes, it's possible S1 had a bunch of items in the pipeline before it ended and was starving S2 of emitting its events the whole time.

A better implementation would round robin both streams to keep it fair, but I was going for speed. Plus, for the purposes of welcome messages it will work ok, since the broadcast stream are local events and we never expect a huge amount of welcome messages at once

Reply:
some improvements & unit tests: https://github.com/xmtp/libxmtp/pull/2182/commits/47874fc6d9ad20339133c346cb91154d1bd7eae4

yesterdays rendition definitely had a starvation issue